simulator/memory/argv_loader/argv_loader.cpp

```diff
@@ -0,0 +1,77 @@
+/**
+ * argv_loader.cpp: loader of arguments to guest memory
+ * @author Rustem Yunusov <yunusov.rr@phystech.edu>
+ * Copyright 2019 MIPT-MIPS iLab project
+ */
+
+#include "argv_loader.h"
+#include <infra/exception.h>
+#include <cstring>
+
+ArgvLoader::ArgvLoader( const char* const* argv, const char* const* envp)
+        : argc( argv ? count_argc( argv) : 0)
+        , argv( argv)
+        , envp( envp)
+        , offset( 0)
+        , envp_offset( 0)
+{
+    if ( !argc)
+        throw InvalidArgs( "argc == 0");
+
+    if ( !argv)
+        throw InvalidArgs( "argv == nullptr");
```

Comment:
We are trying to avoid defensive programming, i.e. we assume that callers always provide meaningful input, and the meainingfulness is guranteed by unit tests, so I don't think we need these checks. Furthermore, you cannot check that the `argv` points to allocated memory anyway.